enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,384 @@
+component and MicroShift.
+
+### Workflow Description
+
+> _Disclaimer:_ Workflows described here illustrate manual processes.
+
+Assuming from now on MicroShift has already been installed in a R4E/RHEL based
+system and there has been some kind of issue.
+
+1. MicroShift administrator notices an issue/wants to get a system report.
+1. The administrator will log in to the failing host.
+1. The administrator will perform the debug report creation procedure:
+```bash
+sudo sos report
+```
+
+For the CI, the same procedure applies.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+There are three main risks:
+1. Upgrades and incompatible changes.
+1. `sos` release cadence.
+1. Extra required disk space.
+
+#### Upgrades
+When upgrading a MicroShift system we may need to upgrade just MicroShift, or
+the OS as well. If it is the latter then `sos` will be updated before
+MicroShift, therefore we must ensure backwards compatibility with the plugin.
+
+If it is the former then we need to ensure newer MicroShift versions do not
+introduce incompatible changes with the reporting capabilities of the `sos`
+plugin. This may also be seen from a different perspective: the plugin should
+make no assumptions about names or locations of components whenever possible.
+In other words, the plugin must be written in the most generic way that still
+allows retrieval of relevant information.
+
+Still, if incompatible changes are introduced then we need to upgrade `sos` as
+well.
+
+#### Sos release cadence
+`sos` release cadence has been shortened to 4 weeks starting on Feb 1st 2023.
+This change has been introduced to allow faster support of new components and
+features. Since MicroShift follows the same release calendar than OpenShift,
+`sos` is faster.
+
+New versions of `sos` will simply slot in to the next regular release. If there
+is an urgent need a hotfix can be made between releases.
+
+#### Extra disk space
+Adding a plugin to `sos` and not relying on MicroShift binary means it needs
+the `oc` or `kubectl` binary installed. This package is available in the RHEL
+repos, but it comes at an extra cost of 125MB once installed.
+
+If we compare it against the minimum disk size required (which is 8GB at the
+time of this writing), it represents an extra 1.5%.
+
+### Drawbacks
+Discussed in risks and mitigations.
+
+## Design Details
+Refer to `sos` [documentation](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin)
+for how to create plugins.
+
+The file structure should ressemble that of must-gather.
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
+MicroShift only tests should be added to verify the output structure: dirs and
+files. The contents may be checked only at the file size level.
+
+An e2e test should test the command exits successfully and certain contents are
+always present.
```

Comment:
Where will that test run? Do we need to add a CI job to the sos repo to ensure compatibility across MicroShift releases?

Reply:
Can we also add a "non-happy path" test please? Keep in mind, this should run in troubled situations. So I suggest a test where we deliberately break something (e.g. create an illegal config, causing CRIO not to start, misconfigure the network causing OVN-K to fail, corrupt container images, whatever). And then make sure sos does run to an end and captures the needed diag info?

Reply:
Done

Reply:
sos CI jobs run in GCE. If we're able to deploy a microshift environment to GCE, we can certainly add a test job to sos. My only question/concern would be implementing a way for us to easily make sure we're always deploying the latest version of microshift there.

Reply:
All MicroShift CI jobs are currently running on GCP, so it should be easy to replicate. We would only need to work out image promotion to public repositories, like quay.

Reply:
What do you think about having a job like the rebase, but for `sos`? Install microshift, install latest version of `sos`, check output and return code, and make a PR to update the dependency if successful.

Reply:
I like the test you describe but it's not clear what the PR is for. What kinds of things would be in that PR?

Reply:
Updating the sos version dependency for the RPM spec. The idea would be to always pick the latest version available.

Reply:
Yes, this is part of the rpm spec. Forward compatibility here is key, as both products have different release cycles.